src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java

```diff
@@ -89,28 +91,86 @@ public void startStreaming(
+            private int numResponses;
 
             @Override
             public void onNext(Vtgate.VStreamResponse response) {
-
                 LOGGER.debug("Received {} VEvents in the VStreamResponse:",
                         response.getEventsCount());
-                for (VEvent vEvent : response.getEventsList()) {
-                    LOGGER.debug("VEvent: {}", vEvent);
+                boolean sendNow = false;
+                for (VEvent event : response.getEventsList()) {
+                    LOGGER.debug("VEvent: {}", event);
+                    switch (event.getType()) {
+                        case ROW:
+                            numOfRowEvents++;
+                            break;
+                        case VGTID:
+                            // We always use the latest VGTID if any.
+                            newVgtid = Vgtid.of(event.getVgtid());
+                            break;
+                        case BEGIN:
+                            // We should only see BEGIN before seeing COMMIT.
+                            if (commitEventSeen) {
+                                LOGGER.error("Receive BEGIN event after receiving COMMIT event");
+                            }
+                            if (beginEventSeen) {
+                                LOGGER.error("Received duplicate BEGIN events");
+                            }
+                            beginEventSeen = true;
+                            break;
+                        case COMMIT:
+                            // We should only see COMMIT after seeing BEGIN.
+                            if (!beginEventSeen) {
+                                LOGGER.error("Receive COMMIT event before receiving BEGIN event");
+                            }
+                            if (commitEventSeen) {
+                                LOGGER.error("Received duplicate COMMIT events");
+                            }
+                            commitEventSeen = true;
+                            break;
+                        case DDL:
+                        case OTHER:
+                            // If receiving DDL and OTHER, process them immediately to rotate vgtid in offset.
+                            // For example, the response can be:
+                            // [VGTID, DDL]. This is an DDL event.
+                            // [VGTID, OTHER]. This is the first response if "current" is used as starting gtid.
+                            sendNow = true;
+                            break;
+                    }
+                    bufferedEvents.add(event);
+                }
+
+                numResponses++;
+
+                // We only proceed when we receive a complete transaction after seeing both BEGIN and COMMIT events,
+                // OR if sendNow flag is true (meaning we should process buffered events immediately).
+                if ((!beginEventSeen || !commitEventSeen) && !sendNow) {
```

Comment:
This seems to be making the assumption that transactions can be sent via multiple parts, and we are being defensive here. I wonder for cases like: `begin1, row1, commit1, begin2`, we would send them out, but the second begin event would still send out with wrong gtid (it will use the gtid from txn1 in this case iiuc) - basically VEvents are ordered but can be split arbitrarily.

Do we need to break the for-loop on on line 106 on commit/ddl/other events and rotate `bufferedEvents`?

Reply:
Yeah. We are not just being defensive. Transactions CAN be split into multiple parts if each part exceeds 250KB (by default), which is the case we encountered.
However, they will not be arbitrarily split. They will only split between BEGIN and COMMIT, therefore `BEGIN1, ROW1, COMMIT1, BEGIN2` should exist as per VStream source code, because right upon COMMIT1, it must send out in its own response.

Reply:
> Transactions CAN be split into multiple parts if each part exceeds 250KB (by default), which is the case we encountered.

For my own understanding, is this limitation on Vitess or MySQL?

Reply:
It's in Vitess: https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/vstreamer/packet_size.go#L27